src/gc/impl/conservative/gc.d

```diff
@@ -1665,8 +1679,10 @@ struct Gcx
         // Return next item from free list
         bucket[bin] = (cast(List*)p).next;
         auto pool = (cast(List*)p).pool;
+
         auto biti = (p - pool.baseAddr) >> pool.shiftBy;
         assert(pool.freebits.test(biti));
+        pool.mark.set(biti); // be sure that the child is aware of the page being used
```

Comment:
`page`->`bin`. This only needs to be executed while a collection is running in the background.

If mark is shared memory between processes, this needs to be an atomic operation (setLocked). Otherwise this might clear bits set concurrently by the other process. But it will also require the background scanning to use atomic operations.

Reply:
Right, I only now see the code commented here is in `smallAlloc()` and just the case bits could be updated while the mark process is also updating them :-/
I wonder if the set operation isn't just always atomic on x86. That would explain why it just works...

Reply:
Another way to deal with this that will decrease concurrency is to just wait for the forked process in this case and on allow allocating new pools while the mark is running (which sucks a bit), or fulfil allocations when the concurrent mark is running with new pools (marking somehow that a pool is new and hidden to the running mark process).

Reply:
IIRC parallel marking didn't work without the atomic operations. These are also a bit more expensive, running the mark function with single threaded, but with atomics, is slower than the non-atomic version.

An alternative might be to keep track of what changed with respect to the freebits between the snapshot and its state when sweep is called. Maybe mark bits don't have to be set at all during allocation.

Reply:
I'm still trying to get around why this have never been a problem in tango runtime with CDGC...

Reply:
> An alternative might be to keep track of what changed with respect to the freebits between the snapshot and its state when sweep is called. Maybe mark bits don't have to be set at all during allocation.

Yeah, that sounds like a good idea, freebits shouldn't be changed until the sweep is done and by then the mark process is already gone.

Reply:
> I'm still trying to get around why this have never been a problem in tango runtime with CDGC...

Allocating during the forked scan is quite different from the parallel marking. If the collection is triggered by a small allocation, all free pages in SmallObjectPools are used and there are probably not that many objects on the free lists anymore (though that might differ with the application). That means that new allocations most often go into new pools which are never scanned by the fork.

IIRC your long running applications also don't allocate that much after the startup phase, so an allocation hitting the same mark-bits as whats currently modifed by the fork is pretty rare.

Reply:
> Yeah, that sounds like a good idea, freebits shouldn't be changed until the sweep is done and by then the mark process is already gone.

freebits are now updated by GC.free and GC.malloc, so I guess a combination of its state when sweep is executing and the copy in the forked memory needs to be used to determine whether an object must be collected.